windows/src/engine/inst-cpp/insthelper.cpp

```diff
@@ -0,0 +1,211 @@
+        DWORD errorCode = GetLastError();
+        std::wstring error =
+            L"Keyman Engine failed to set permissions on shared data in CreateDir: " + std::to_wstring(errorCode);
+        MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+        return errorCode;
+      }
+    }
+
+    // Create file handle
+    hFile = CreateFile(path, READ_CONTROL | WRITE_DAC, 0, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    if (hFile == INVALID_HANDLE_VALUE) {
+      DWORD errorCode    = GetLastError();
+      std::wstring error = L"Keyman Engine failed to set permissions on shared data in CreateFile: " + std::to_wstring(errorCode);
+      MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+      return errorCode;
+    }
+
+    // Set permission on shared data
+    EXPLICIT_ACCESS ea      = {0};
+    ea.grfAccessPermissions = GENERIC_READ | GENERIC_EXECUTE;
+    ea.grfAccessMode        = SET_ACCESS;
+    ea.grfInheritance       = SUB_CONTAINERS_AND_OBJECTS_INHERIT;
+    ea.Trustee.TrusteeForm  = TRUSTEE_IS_NAME;
+    ea.Trustee.ptstrName    = (LPWCH)L"ALL APPLICATION PACKAGES";
+
+    PACL pOldDACL = NULL;
+
+    DWORD result = SetEntriesInAcl(1, &ea, NULL, &pOldDACL);
+    if (result != ERROR_SUCCESS) {
+      CloseHandle(hFile);
+      std::wstring error =
+          L"Keyman Engine failed to set permissions on shared data in GrantPermission: " + std::to_wstring(result);
+      MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+      if (pOldDACL) {
+        LocalFree(pOldDACL);
+      }
+      return result;
+    }
+
+    result = SetNamedSecurityInfo(path, SE_FILE_OBJECT, DACL_SECURITY_INFORMATION, NULL, NULL, pOldDACL, NULL);
+    if (result != ERROR_SUCCESS) {
+      std::wstring error = L"Keyman Engine failed to apply DACL to shared data folder: " + std::to_wstring(result);
+      MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+    }
+
+    if (pOldDACL) {
+      LocalFree(pOldDACL);
+    }
+
+    CloseHandle(hFile);
+  }
+
+  CoTaskMemFree(path);
+  return ERROR_SUCCESS;
+}
+
+// Define the CLS ID for KMTip Text Service
+DEFINE_GUID(c_clsidKMTipTextService, 0xFE0420F1, 0x38D1, 0x4B4C, 0x96, 0xBF, 0xE7, 0xE2, 0x0A, 0x74, 0xCF, 0xB7);
+
+// Define the CLASS_TF_InputProcessorProfiles constant
+DEFINE_GUID(CLASS_TF_InputProcessorProfiles, 0x33C53A50, 0xF456, 0x4884, 0xB0, 0x49, 0x85, 0xFD, 0x64, 0x3E, 0xCF, 0xED);
+
+DEFINE_GUID(IID_ITfInputProcessorProfiles, 0x1F02B6C5, 0x7842, 0x4EE6, 0x8A, 0x0B, 0x9A, 0x24, 0x18, 0x3A, 0x95, 0xCA);
+
+const DWORD ILOT_UNINSTALL = 1;
+
+std::wstring
+GuidToWString(const GUID& guid) {
+  wchar_t guidCStr[40] = {0};
+  int length           = StringFromGUID2(guid, guidCStr, 40);
+
+  if (length > 0) {
+    // Successfully converted GUID to wstring
+    return std::wstring(guidCStr);
+  } else {
+    // Handle the error, for example, return an empty wstring
+    return std::wstring();
+  }
```

Comment:
We shouldn't be returning an empty string on failure because that's going to mask an error which causes problems further up the call stack.

(That said, StringFromGUID2 should never fail, because its only failure case is if the buffer is too small, which is not the case here.)